src/core/net/dns_client.cpp

```diff
@@ -1106,6 +1191,121 @@ void Client::HandleTimer(void)
+}
+
+void Client::HandleTcpSendDoneCallback(otTcpEndpoint *aEndpoint, otLinkedBuffer *aData)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    OT_UNUSED_VARIABLE(aData);
+}
+
+void Client::HandleTcpReceiveAvailable(otTcpEndpoint *aEndpoint,
+                                       size_t         aBytesAvailable,
+                                       bool           aEndOfStream,
+                                       size_t         aBytesRemaining)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    OT_UNUSED_VARIABLE(aBytesRemaining);
+
+    const otLinkedBuffer *data;
+    size_t                totalReceived = 0;
+    Message              *message       = nullptr;
+    Response              response;
+    QueryType             type;
+    Error                 responseError;
+    uint16_t              length = 0;
+
+    if (aEndOfStream)
+    {
+        // Cleanup is done in dissconnected callback.
+        mTcpTimer.Stop();
+    }
+
+    VerifyOrExit(aBytesAvailable != 0);
+    mEndpoint.ReceiveByReference(data);
+    message = mSocket.NewMessage(0);
+    VerifyOrExit(message != nullptr);
+
+    if (data != nullptr)
+    {
+        totalReceived += data->mLength;
+        message->AppendBytes(data->mData + kTcpQueryLenSize, data->mLength - kTcpQueryLenSize);
+        length = ReadUint16(data->mData);
+        if (length > OPENTHREAD_CONFIG_DNS_OVER_TCP_QUERY_MAX_SIZE)
+        {
+            LogWarn("Dns query over wasn't received message too big.");
+            responseError = kErrorNoBufs;
+        }
+    }
+    mEndpoint.CommitReceive(totalReceived, 0);
+    response.mInstance = &Get<Instance>();
+    response.mMessage  = message;
+
+    SuccessOrExit(ParseResponse(response, type, responseError));
+    FinalizeQuery(response, type, responseError);
+
+exit:
+    if (message != nullptr)
+    {
+        message->Free();
+    }
+    return;
+}
+#include <zephyr/kernel.h>
+void Client::HandleTcpReceiveAvailableCallback(otTcpEndpoint *aEndpoint,
+                                               size_t         aBytesAvailable,
+                                               bool           aEndOfStream,
+                                               size_t         aBytesRemaining)
+{
+    LogDebg("Dns tcp connection receive available callback.\n");
+    static_cast<Client *>(otTcpEndpointGetContext(aEndpoint))
+        ->HandleTcpReceiveAvailable(aEndpoint, aBytesAvailable, aEndOfStream, aBytesRemaining);
+}
+
+void Client::HandleTcpDisconnected(otTcpEndpoint *aEndpoint, otTcpDisconnectedReason aReason)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    Query *query;
+
+    LogDebg("Dns TCP disconnected reason: %d\n", aReason);
+    mEndpoint.Deinitialize();
+    mTcpInitialized = false;
+    VerifyOrExit(aReason != OT_TCP_DISCONNECTED_REASON_NORMAL);
+
+    // Abort queries in case of connection failures
+    while ((query = mQueries.GetHead()) != nullptr)
+    {
+        FinalizeQuery(*query, kErrorAbort);
+    }
```

Comment:
I think we should only finalize the queries that are using TCP on TCP close.
There may be other queries that are using UDP which can continue.

Reply:
fixed